src/main/java/io/rebloom/client/Client.java

```diff
@@ -358,4 +372,262 @@ private Connection sendCommand(Jedis conn, ProtocolCommand command, byte[]... ar
+
+      List<byte[]> fullArgs = new ArrayList<byte[]>();
+      fullArgs.add(SafeEncoder.encode(key));
+      fullArgs.addAll(options.asListOfByteArrays());
+      fullArgs.add(SafeEncoder.encode("ITEMS"));
+      fullArgs.addAll(Arrays.asList(SafeEncoder.encodeMany(items)));
+
+      return sendCommand(conn, CuckooCommand.INSERT, fullArgs) //
+          .getIntegerMultiBulkReply() //
+          .stream() //
+          .map(s -> s > 0) //
+          .collect(Collectors.toList());
+    }
+  }
+
+  @Override
+  public List<Boolean> cfInsertNx(String key, String... items) {
+    try (Jedis conn = _conn()) {
+
+      List<byte[]> fullArgs = new ArrayList<byte[]>();
+      fullArgs.add(SafeEncoder.encode(key));
+      fullArgs.add(SafeEncoder.encode("ITEMS"));
+      fullArgs.addAll(Arrays.asList(SafeEncoder.encodeMany(items)));
+
+      return sendCommand(conn, CuckooCommand.INSERTNX, fullArgs) //
+          .getIntegerMultiBulkReply() //
+          .stream() //
+          .map(s -> s > 0) //
+          .collect(Collectors.toList());
+    }
+  }
+
+  @Override
+  public List<Boolean> cfInsertNx(String key, CFInsertOptions options, String... items) {
+    try (Jedis conn = _conn()) {
+
+      List<byte[]> fullArgs = new ArrayList<byte[]>();
+      fullArgs.add(SafeEncoder.encode(key));
+      fullArgs.addAll(options.asListOfByteArrays());
+      fullArgs.add(SafeEncoder.encode("ITEMS"));
+      fullArgs.addAll(Arrays.asList(SafeEncoder.encodeMany(items)));
+
+      return sendCommand(conn, CuckooCommand.INSERTNX, fullArgs) //
+          .getIntegerMultiBulkReply() //
+          .stream() //
+          .map(s -> s > 0) //
+          .collect(Collectors.toList());
+    }
+  }
+
+  @Override
+  public boolean cfExists(String key, String item) {
+    try (Jedis conn = _conn()) {
+      return sendCommand(conn, CuckooCommand.EXIST, //
+          SafeEncoder.encode(key), //
+          SafeEncoder.encode(item) //
+      ).getIntegerReply() == 1;
+    }
+  }
+
+  @Override
+  public boolean cfDel(String key, String item) {
+    try (Jedis conn = _conn()) {
+      return sendCommand(conn, CuckooCommand.DEL, //
+          SafeEncoder.encode(key), //
+          SafeEncoder.encode(item) //
+      ).getIntegerReply() == 1;
+    }
+  }
+
+  @Override
+  public long cfCount(String key, String item) {
+    try (Jedis conn = _conn()) {
+      return sendCommand(conn, CuckooCommand.COUNT, //
+          SafeEncoder.encode(key), //
+          SafeEncoder.encode(item) //
+      ).getIntegerReply();
+    }
+  }
+
+  @Override
+  public Map.Entry<Long, byte[]> cfScanDump(String key, Long iterator) {
+    try (Jedis conn = _conn()) {
+      List<Object> response = sendCommand(conn, CuckooCommand.SCANDUMP, //
+          SafeEncoder.encode(key), //
+          Protocol.toByteArray(iterator) //
+      ).getObjectMultiBulkReply();
+      
+      return new AbstractMap.SimpleImmutableEntry<Long, byte[]>(((Long) response.get(0)).longValue(), (byte[]) response.get(1));
```

Comment:
**`Protocol.toByteArray(long value)` considers `value` as general (signed) long.** You have to handle unsigned long by yourself. Same goes for `getObjectMultiBulkReply()`.

Reply:
@sazzad16 how about doing something like:
```
  @Override
  public Map.Entry<Long, byte[]> cfScanDump(String key, long iterator) {
    try (Jedis conn = _conn()) {
      List<Object> response = sendCommand(conn, CuckooCommand.SCANDUMP, //
          SafeEncoder.encode(key), //
          longToByteArray(iterator) //
      ).getObjectMultiBulkReply();
      
      return new AbstractMap.SimpleImmutableEntry<Long, byte[]>(((Long) response.get(0)).longValue(), (byte[]) response.get(1));
    }
  }
```
where `longToByteArray` is:
```
  private byte[] longToByteArray(long l) {
    String asUnsignedString = Long.toUnsignedString(l);
    return SafeEncoder.encode(asUnsignedString);
  }  
```
I tried the above and the tests pass. Perhaps a mock tests with a unsigned long byte payload is required?  But more importantly I wonder what is a good solution for the `getMultiBulkReply`. I see that in Jedis is boils down to the `process` method in the `Protocol` class:
```
  private static Object process(final RedisInputStream is) {
    final byte b = is.readByte();
    switch (b) {
    case PLUS_BYTE:
      return processStatusCodeReply(is);
    case DOLLAR_BYTE:
      return processBulkReply(is);
    case ASTERISK_BYTE:
      return processMultiBulkReply(is);
    case COLON_BYTE:
      return processInteger(is);
    case MINUS_BYTE:
      processError(is);
      return null;
    default:
      throw new JedisConnectionException("Unknown reply: " + (char) b);
    }
  }
 ```

What I need to understand is if you are saying that we need to handle the `COLON_BYTE` case above which calls `processInteger`-> `readLongCrLf`? And do a `byte[]` to `String` and then a `Long.parseUnsignedLong(s)` to get a long? Any suggestions would be welcomed.

Reply:
@bsbodden Please disregard my change request regarding this issue.

In RedisBloom `CF.SCANDUMP` implementation https://github.com/RedisBloom/RedisBloom/blob/605086fc92f820257e45004440533060be90c23e/src/rebloom.c#L771-L816 , an unsigned 64-bit number is replied as iterator. See `long long pos;` and the usages of `RedisModule_ReplyWithLongLong()` there. That's why `((Long) response.get(0))` part in your Java code passing. So we can't handle the case I'm talking about even if we want to because it won't be backed by RedisBloom itself. It's also possible that CF.SCANDUMP doesn't even generate the case!

I got confused because in similar cases, Redis replies with a bulk string, instead of a number in any format.

Sorry to move you back and forth.

Reply:
No problem at all! The back and forth is how we make the code better!